keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,1114 @@
+    status: "True"
+    type: ParamsReady
+```
+##### Status Conditions
+The PodNetwork object will use following conditions:
+* **Ready** - indicates that the PodNetwork object is correct (validated) and
+all other conditions are set to “true”. This condition will switch back to
+“false” if ParamsReady condition is “false”. Pods cannot be attached to a
+PodNetwork that is not Ready. This condition does not indicate readiness of
+specific PodNetwork on a per Node-basis. Following are the error reasons for
+this condition:
+
+| Reason name              | Description                                                                                                                      |
+|--------------------------|----------------------------------------------------------------------------------------------------------------------------------|
+| ParamsNotReady           | The ParamsReady condition is not present or has “false” value. This can only happen when the “parametersRefs” field has a value. |
+| AdministrativelyDisabled | The PodNetwork’s Enabled field is set to false.                                                                                  |
+
+* **ParamsReady** - indicates that the objects specified in the “parametersRefs”
+field are ready for use. The implementation (effectively the owner of the
+specified objects) is responsible for setting this condition to “true” after
+having performed whatever checks the implementation requires (such as validating
+the CRs and checking that the network they reference is ready). The “Ready”
+condition is dependent on the value of this condition when the “parametersRefs”
+field is not empty. The available “reasons” for this condition are implementation
+specific. When multiple references are provided in the “parametersRefs” field,
+it is implementation responsibility to provide accurate status for all the listed
+objects using this one condition.
+
+The conditions life-cycle will be handled by the PodNetwork controller described
+below.
+
+##### Provider
+The provider field gives the ability to uniquely identify what implementation
+(provider) is going to handle specific instances of PodNetwork objects. The value
+has to be in the form of an url. It is the implementer’s decision on how this field
+is going to be leveraged. They will decide how they behave when this field is
+empty and what specific value they are going to respect. This will dictate if a
+specific implementation can co-exist with other ones in the same cluster.
+
+We have considered using classes (similar to GatewayClass etc.), but we do not
+expect any extra action for a PodNetwork to take place for specific implementation.
+PodNetwork already points to a custom resource, which implementation can hook on
+for any specific extra configuration. At this point of time we consider using
+classes as overkill, and in future if such need arises, we can reuse the provider
+field for that purpose.
```

Comment:
So, I think your model of how this works is that there's a single component in the cluster responsible for all PodNetworks.

But another model would be to allow multiple PodNetwork-managing components in the cluster, each supporting different functionality, and managing only the PodNetworks of their own PodNetworkClass. For example, there might be separate PodNetworkClasses implementing each of the User Stories (though there could also be "heavyweight" multi-use-case PodNetworkClasses like the implied universal PodNetworkClass currently envisioned by the KEP).

I think the PodNetworkClass model would lead to more reused code and more common functionality across cluster types. Given the model of the current KEP, it's possible (indeed, likely) that, say, GKE and OpenShift could both provide ways of implementing all of the User Stories, but those ways would be entirely incompatible, involving different `Provider` names and `ParametersRefs` of completely different CRD types. Whereas with the PodNetworkClass model, GKE and OpenShift might each have some PodNetworkClasses of their own, but could also make use of "generic" ones (possibly ones written by users themselves).

Reply:
The multiple implementation in cluster suppose to be handled by the provider field, assuming the implementations support that field, and this makes the interoperability between implementation dependent on each other.

With class this is enforced if we make that field mandatory. Will bring this up in next sync meeting and talk it thru.